### ISSUE TITLE
Bundle the react-native packager

### ISSUE THREAD

**Issue by DogLooksGood** [status-im/clj-rn#1]
Thank you for the great work and efforts that makes the complex `re-natal` into a simple handy tool.

Now there's a separate project for React Native packager.
https://github.com/facebook/metro

It will be more convenient if `clj-rn` can start packager server and simplify basic dev workflow.

The desired workflow may like:
```
$ npx react-native install    # only once if there's no native code change
$ clj -R repl -m clj-rn.main watch
```

It has to be clarify how deployment of the app itself is done as it is sometimes necessary to update it (and it's at least needed once).

Specifically both `react-native start` and `react-native run-*` should be replaced and made irrelevant.

Some links:

* https://github.com/facebook/react-native/tree/master/local-cli/runAndroid
* https://github.com/facebook/react-native/tree/master/local-cli/runIOS
* https://github.com/facebook/react-native/blob/master/local-cli/cliEntry.js

**Comment by jeluard** [comment]
Makes a lot of sense!

We probably want to extract the `figwheel-bridge` part in a dedicated issue so that PR can be more atomic.

**Comment by status-open-bounty** [comment]
Balance: 0 ETH
Tokens: SNT: 5000.00
Contract address: [0x1a98b995c4950b030bb60db049829bafce54a53e](https://etherscan.io/address/0x1a98b995c4950b030bb60db049829bafce54a53e)
Network: Mainnet
Paid to: endenwer
Visit [https://openbounty.status.im](https://openbounty.status.im) to learn more.

**Comment by jeluard** [comment]
@DogLooksGood To make this PR more atomic I introduced 2 new PRs: #2 and #3 and simplified this one. Let me know if that makes sense :)

**Comment by DogLooksGood** [comment]
@jeluard Yes, these make sense.

**Comment by t1MZXxBn6pAAtTqKTy5UotdeMqBw2CfGrUF** [comment]
Makes sense.

**Comment by endenwer** [comment]
Hey @jeluard. As I understand there is a need for two commands: `clj -R repl -m clj-rn.main watch` and `clj -R repl -m clj-rn.main run-*` which replace `react-native start` and `react-native run-*` respectively. Is that right? I can start working on it.

**Comment by jeluard** [comment]
Hey @endenwer ! So the `watch` one would ideally be self-sufficient so that it's the only command you need to use to have your dev workflow ready. So extending `main` one with all the things done by `react-native start`. All as clojure fns and probably reusing metro JS code.

Then the other one is indeed replacing `run-*` commands (also available as fn, so it can be triggered in the REPL).
It would be great to figure out how to abstract from ios/android based on what the associated `watch` is serving.

Hope it helps!

**Comment by endenwer** [comment]
@jeluard For development you need to run three commands:
1. React native server: `react-native start` or `clj -R repl -m clj-rn.main watch`;
2. Figwheel or repl with figweel. Now you can do it by using task `figwheel` in status-react `build.clj` which use `rebuild-index` and `enable-source-maps` from this repo;
3. Bundle android or ios build and run it: `react-native run-*` or `clj -R repl -m clj-rn.main run-* `.

> It would be great to figure out how to abstract from ios/android based on what the associated watch is serving.

Maybe just use the same defaults as here https://github.com/facebook/react-native/blob/54942746d4037e1153e14fcfc95e4edc772d296a/local-cli/core/index.js#L76

**Comment by jeluard** [comment]
Right we want both 1 and 2 to be combined. No reason to have them separate.

For the `run` one, it should be agnostic and just pick up the right thing based on what the first command is serving.

**Comment by jeluard** [comment]
@endenwer Does that sound possible?

**Comment by endenwer** [comment]
@jeluard 
> Right we want both 1 and 2 to be combined. No reason to have them separate.

Actually there is a reason to start it separately. I start figwheel within repl which I run in emacs. At least there should be an option to run it separately. 

I suggest to add a task `start-server` and then command `watch` will run `start-server` and `figwheel`. Figwheel command can be moved from `build.clj` in status-react. 
Also `watch` will write build name to some local file and `run` will use it to run the same build.

TODO:
1. Add task `start-server` which starts metro server
2. Move task `figwheel` from status-react
3. Add task `watch` which run tasks 1 and 2. This task will require build ids.
3. Add task `run` which run all builds from previous the task. 

After all changes dev workflow will be like this:
```
clj -R repl -m clj-rn.main watch android
clj -R repl -m clj-rn.main run
```

**Comment by jeluard** [comment]
> At least there should be an option to run it separately.

Good point

Your suggestions look perfect. As much as possible this library should be self-contained and allow simple RN dev workflow!

**Comment by jeluard** [comment]
Acually I am not even sure we need a separate task for run at all. Maybe it would be simpler to have a REPL command only.

**Comment by endenwer** [comment]
@jeluard I don't think it is possible to run build from repl. When you start figwheel you see message `Prompt will show when Figwheel connects to your application`. Application should connect to figwheel to start repl so you can't use it before starting `run` task.

If everything is ok then I start working on it.

**Comment by jeluard** [comment]
Hmm indeed there is a chicken and egg problem, at least for the first deploy on mobile.

Please go ahead!

**Comment by DogLooksGood** [comment]
maybe it doesn't matter to combine 1 and 2 into one command for using with emacs. 

shadow-cljs has this approach, just start with `npx shadow-cljs watch ios`(with cider just `cider-jack-in`), it will start metro first, and its stdout will be redirect to repl. then a repl will be fired up. this is really convenient, you don't have to open multiple terminals for these processes.

**Comment by endenwer** [comment]
That is exactly what I want to do. `watch` command will start metro and figwheel. Then figwheel start repl. You can actually connect to this repl from emacs, but in some cases you may need to start repl differently and this is why there should be ability to run it separately.

**Comment by endenwer** [comment]
Hey @jeluard I have good news. I can do just one command. But there is a problem with metro, `react-native start` not only start metro server but also start many different middleware https://github.com/facebook/react-native/blob/master/local-cli/server/middleware/MiddlewareManager.js#L47-L55. And also metro can only build application and can't start it and it's about 1000 lines of code in react-native. So I think to use react-native instead of metro. Is that ok or you want to rewrite all the code from react-native?

**Comment by jeluard** [comment]
Awesome!
Just to clarify, by using `react-native` you mean using react-native JS lib? If so sounds good!

**Comment by endenwer** [comment]
I mean using `react-native start` and `react-native run-*`

**Comment by jeluard** [comment]
Shelling out and calling those commands? Are those commands not implemented in JS that could somehow be reused?
I'd rather have direct code dependency and usage.

**Comment by endenwer** [comment]
How can I reuse js in clojure project? If it would be cljs then I would do it, but here I need to start it like `react-native start` or `node path/to/react-native/file`

**Comment by jeluard** [comment]
Right.. Well technically you could do it with Java ScriptEngine (http://www.oracle.com/technetwork/articles/java/jf14-nashorn-2126515.html) but sounds like a giant pain..

Can you give it a try with shell commands to see how it goes?

**Comment by endenwer** [comment]
I want to do it like this:
1. Exec `react-native start`
2. Start figwheel server(without repl)
3. Exec `react-native run-*` for every build
4. Start cljs repl

User need to start only one command `clj -R repl -m clj-rn.main watch android ios`

**Comment by jeluard** [comment]
Ok. I would make step 3. optional.
Let's see how it goes!

**Comment by endenwer** [comment]
awesome, I will make it optional.

**Comment by pablanopete** [comment]
@endenwer sorry for the slow response on bounty payout, could you drop your wallet address here so I can send you the bounty amount Out of Band while we wait on a dashboard fix in Status Open Bounty. 

If you don't feel comfortable putting it here in the comment could you email a ERC20 compatible wallets address to me? chris.hutchinson@status.im

**Comment by endenwer** [comment]
Hey @pablanopete. Thanks for response. Here is my address 0xfa6749538338048312a7c14858be03b96ed3dd11

**Comment by pablanopete** [comment]
Send @endenwer ! Will also be sending the original amount once the SOB dashboard bugs are ironed out; thanks for your patience and contributions

https://etherscan.io/tx/0x2550b7d248ff0731575329dad4edb4c1b15534e3e7740a8c3c7aed905461ea3e
